File: app/src/main/java/com/hiddenramblings/tagmo/eightbit/bluetooth/BluetoothHandler.java
/* ====================================================================
 * Copyright (c) 2012-2022 AbandonedCart.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * For the purpose of this license, the phrase "TagMo labels" shall
 * be used to refer to the labels "8-Bit Dream", "TwistedUmbrella",
 * "TagMo" and "AbandonedCart" and these labels should be considered
 * the equivalent of any usage of the aforementioned phrase.
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All materials mentioning features or use of this software and
 *    redistributions of any form whatsoever must display the following
 *    acknowledgment unless made available by tagged, public "commits":
 *    "This product includes software developed for TagMo by AbandonedCart"
 *
 * 4. The TagMo labels must not be used in any form to endorse or promote
 *    products derived from this software without prior written permission.
 *    For written permission, please contact enderinexiledc@gmail.com
 *
 * 5. Products derived from this software may not be called by the TagMo labels
 *    nor may these labels appear in their names or product information without
 *    prior written permission of AbandonedCart.
 *
 * THIS SOFTWARE IS PROVIDED BY AbandonedCart AND TagMo ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * The license and distribution terms for any publicly available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution license
 * [including the GNU Public License.] Content not subject to these terms is
 * subject to to the terms and conditions of the Apache License, Version 2.0.
 */

package com.hiddenramblings.tagmo.eightbit.bluetooth;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.ActivityResultRegistry;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;

import java.util.Map;

public class BluetoothHandler {

    private BluetoothAdapter mBluetoothAdapter;
    private final BluetoothListener listener;

    ActivityResultLauncher<String[]> onRequestLocationQ;
    ActivityResultLauncher<String> onRequestBackgroundQ;
    ActivityResultLauncher<String[]> onRequestBluetoothS;
    ActivityResultLauncher<Intent> onRequestBluetooth;
    ActivityResultLauncher<String[]> onRequestLocation;

    public BluetoothHandler(Context context, ActivityResultRegistry registry, BluetoothListener listener) {
        onRequestLocationQ = registry.register("LocationQ",
                new ActivityResultContracts.RequestMultiplePermissions(),
                permissions -> { boolean isLocationAvailable = false;
            for (Map.Entry<String,Boolean> entry : permissions.entrySet()) {
                if (entry.getKey().equals(Manifest.permission.ACCESS_FINE_LOCATION)
                        && entry.getValue()) isLocationAvailable = true;
            }
            if (isLocationAvailable) {
                requestBluetooth(context);
            } else {
                listener.onPermissionsFailed();
            }
        });

        onRequestBackgroundQ = registry.register("BackgroundQ",
                new ActivityResultContracts.RequestPermission(), permission -> {});
        
        onRequestBluetoothS = registry.register("BluetoothS",
                new ActivityResultContracts.RequestMultiplePermissions(),
                permissions -> { boolean isBluetoothAvailable = false;
            for (Map.Entry<String,Boolean> entry : permissions.entrySet()) {
                if (entry.getValue()) isBluetoothAvailable = true;
            }
            if (isBluetoothAvailable) {
                BluetoothAdapter mBluetoothAdapter = getBluetoothAdapter(context);
                if (null != mBluetoothAdapter) {
//                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
//                        onRequestBackgroundQ.launch(Manifest.permission.ACCESS_BACKGROUND_LOCATION);
//                    }
                    listener.onAdapterEnabled(mBluetoothAdapter);
                } else {
                    listener.onAdapterMissing();
                }
            } else {
                listener.onAdapterMissing();
            }
        });
        
        onRequestBluetooth = registry.register("Bluetooth",
                new ActivityResultContracts.StartActivityForResult(), result -> {
            BluetoothAdapter mBluetoothAdapter = getBluetoothAdapter(context);
            if (null != mBluetoothAdapter) {
//                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
//                    onRequestBackgroundQ.launch(Manifest.permission.ACCESS_BACKGROUND_LOCATION);
//                }
                listener.onAdapterEnabled(mBluetoothAdapter);
            } else {
                listener.onAdapterMissing();
            }
        });
        
        onRequestLocation = registry.register("Location", 
                new ActivityResultContracts.RequestMultiplePermissions(), 
                permissions -> { boolean isLocationAvailable = false;
            for (Map.Entry<String,Boolean> entry : permissions.entrySet()) {
                if (entry.getValue()) isLocationAvailable = true;
            }
            if (isLocationAvailable) {
                BluetoothAdapter mBluetoothAdapter = getBluetoothAdapter(context);
                if (null != mBluetoothAdapter)
                    listener.onAdapterEnabled(mBluetoothAdapter);
                else
                    onRequestBluetooth.launch(new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE));
            } else {
                listener.onPermissionsFailed();
            }
        });

        this.listener = listener;
    }

    private void requestBluetooth(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            final String[] PERMISSIONS_BLUETOOTH = {
                    Manifest.permission.BLUETOOTH_CONNECT,
                    Manifest.permission.BLUETOOTH_SCAN
            };
            onRequestBluetoothS.launch(PERMISSIONS_BLUETOOTH);
        } else {
            BluetoothAdapter mBluetoothAdapter = getBluetoothAdapter(context);
            if (null != mBluetoothAdapter) {
//                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
//                    onRequestBackgroundQ.launch(Manifest.permission.ACCESS_BACKGROUND_LOCATION);
//                }
                listener.onAdapterEnabled(mBluetoothAdapter);
            } else {
                onRequestBluetooth.launch(new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE));
            }
        }
    }

    public void requestPermissions(Activity context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            BluetoothAdapter mBluetoothAdapter = getBluetoothAdapter(context);
            if (null != mBluetoothAdapter)
                listener.onAdapterEnabled(mBluetoothAdapter);
            else
                onRequestBluetooth.launch(new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE));
        } else {
            if (ActivityCompat.shouldShowRequestPermissionRationale(
                    context, Manifest.permission.ACCESS_FINE_LOCATION
            )) {
                new AlertDialog.Builder(context)
                        .setMessage(R.string.location_disclosure)
                        .setCancelable(false)
                        .setPositiveButton(R.string.accept, (dialog, which) -> {
                            dialog.dismiss();
                            final String[] PERMISSIONS_LOCATION = {
                                    Manifest.permission.ACCESS_FINE_LOCATION,
                                    Manifest.permission.ACCESS_COARSE_LOCATION
                            };
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                                onRequestLocationQ.launch(PERMISSIONS_LOCATION);
                            } else {
                                onRequestLocation.launch(PERMISSIONS_LOCATION);
                            }
                        }).setNegativeButton(R.string.deny, (dialog, which) ->
                                listener.onPermissionsFailed()).show();
            } else if (ContextCompat.checkSelfPermission(
                    context, Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED) {
                requestBluetooth(context);
            } else {
                if (TagMo.isGooglePlay()) {
                    new AlertDialog.Builder(context)
                            .setMessage(R.string.location_disclosure)
                            .setCancelable(false)
                            .setPositiveButton(R.string.accept, (dialog, which) -> {
                                dialog.dismiss();
                                final String[] PERMISSIONS_LOCATION = {
                                        Manifest.permission.ACCESS_FINE_LOCATION,
                                        Manifest.permission.ACCESS_COARSE_LOCATION
                                };
                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                                    onRequestLocationQ.launch(PERMISSIONS_LOCATION);
                                } else {
                                    onRequestLocation.launch(PERMISSIONS_LOCATION);
                                }
                            }).setNegativeButton(R.string.deny, (dialog, which) ->
                                    listener.onPermissionsFailed()).show();
                } else {
                    final String[] PERMISSIONS_LOCATION = {
                            Manifest.permission.ACCESS_FINE_LOCATION,
                            Manifest.permission.ACCESS_COARSE_LOCATION
                    };
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                        onRequestLocationQ.launch(PERMISSIONS_LOCATION);
                    } else {
                        onRequestLocation.launch(PERMISSIONS_LOCATION);
                    }
                }
            }
        }
    }

    @SuppressLint("MissingPermission")
    public BluetoothAdapter getBluetoothAdapter(Context context) {
        if (null == context) return null;
        if (null != mBluetoothAdapter) {
            if (!mBluetoothAdapter.isEnabled()) mBluetoothAdapter.enable();
            return mBluetoothAdapter;
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                mBluetoothAdapter = ((BluetoothManager) context.getSystemService(
                        Context.BLUETOOTH_SERVICE
                )).getAdapter();
            } else {
                mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
            }
            if (null != mBluetoothAdapter) {
                if (!mBluetoothAdapter.isEnabled()) mBluetoothAdapter.enable();
                return mBluetoothAdapter;
            }
        }
        return null;
    }

    public interface BluetoothListener {
        void onPermissionsFailed();
        void onAdapterMissing();
        void onAdapterEnabled(BluetoothAdapter adapter);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/bluetooth/BluetoothHandler.kt
/*
 * ====================================================================
 * Copyright (c) 2021-2023 AbandonedCart.  All rights reserved.
 *
 * https://github.com/AbandonedCart/AbandonedCart/blob/main/LICENSE#L4
 * ====================================================================
 *
 * The license and distribution terms for any publicly available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution license
 * [including the GNU Public License.] Content not subject to these terms is
 * subject to to the terms and conditions of the Apache License, Version 2.0.
 */
package com.hiddenramblings.tagmo.bluetooth

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.provider.Settings
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.ActivityResultRegistry
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.hiddenramblings.tagmo.BuildConfig
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.eightbit.os.Version

class BluetoothHandler(
    context: Context, registry: ActivityResultRegistry, listener: BluetoothListener
) {
    private var mBluetoothAdapter: BluetoothAdapter? = null
    private val listener: BluetoothListener
    private var onRequestLocationQ: ActivityResultLauncher<String>
    private var onRequestBluetoothS: ActivityResultLauncher<Array<String>>
    private var onRequestBluetooth: ActivityResultLauncher<Intent>
    private var onRequestLocation: ActivityResultLauncher<String>
    private var onRequestAdapter: ActivityResultLauncher<Intent>

    init {
        onRequestLocationQ = registry.register(
                "LocationQ", ActivityResultContracts.RequestPermission()
        ) { granted ->
            if (granted) requestBluetooth(context) else listener.onPermissionsFailed()
        }
        onRequestBluetoothS = registry.register(
            "BluetoothS", ActivityResultContracts.RequestMultiplePermissions()
        ) { permissions: Map<String, Boolean> ->
            if (permissions.entries.all { it.value }) {
                val mBluetoothAdapter = getBluetoothAdapter(context)
                mBluetoothAdapter?.let { listener.onAdapterEnabled(it) }
                    ?: listener.onAdapterMissing()
            } else {
                listener.onAdapterMissing()
            }
        }
        onRequestBluetooth = registry.register(
            "Bluetooth", ActivityResultContracts.StartActivityForResult()
        ) {
            val mBluetoothAdapter = getBluetoothAdapter(context)
            mBluetoothAdapter?.let { listener.onAdapterEnabled(it) } ?: listener.onAdapterMissing()
        }
        onRequestLocation = registry.register(
            "Location", ActivityResultContracts.RequestPermission()
        ) { granted ->
            if (granted) {
                val mBluetoothAdapter = getBluetoothAdapter(context)
                mBluetoothAdapter?.let { listener.onAdapterEnabled(it) }
                    ?: onRequestBluetooth.launch(Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE))
            } else {
                listener.onPermissionsFailed()
            }
        }
        onRequestAdapter = registry.register(
            "Adapter", ActivityResultContracts.StartActivityForResult()
        ) { listener.onAdapterRestricted() }
        this.listener = listener
    }

    private fun requestBluetooth(context: Context) {
        if (Version.isSnowCone) {
            onRequestBluetoothS.launch(PERMISSIONS_BLUETOOTH)
        } else {
            val mBluetoothAdapter = getBluetoothAdapter(context)
            mBluetoothAdapter?.let { listener.onAdapterEnabled(it) }
                ?: onRequestBluetooth.launch(Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE))
        }
    }

    private fun showLocationRequest(context: Context) {
        AlertDialog.Builder(context)
            .setMessage(R.string.location_disclosure)
            .setCancelable(false)
            .setPositiveButton(R.string.accept) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                if (Version.isQuinceTart)
                    onRequestLocationQ.launch(PERMISSION_FINE_LOCATION)
                else
                    onRequestLocation.launch(PERMISSION_COURSE_LOCATION)
            }
            .setNegativeButton(R.string.deny) { _: DialogInterface?, _: Int ->
                listener.onPermissionsFailed() }
            .show()
    }

    fun requestPermissions(activity: Activity) {
        when {
            Version.isLowerThan(Build.VERSION_CODES.M) -> {
                val mBluetoothAdapter = getBluetoothAdapter(activity)
                mBluetoothAdapter?.let { listener.onAdapterEnabled(it) }
                    ?: onRequestBluetooth.launch(Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE))
            }
            Version.isSnowCone -> requestBluetooth(activity)
            Version.isQuinceTart -> {
                when {
                    ActivityCompat.shouldShowRequestPermissionRationale(
                        activity, PERMISSION_FINE_LOCATION
                    ) -> {
                        showLocationRequest(activity)
                    }
                    ContextCompat.checkSelfPermission(
                        activity, PERMISSION_FINE_LOCATION
                    ) == PackageManager.PERMISSION_GRANTED -> {
                        requestBluetooth(activity)
                    }
                    else -> {
                        if (BuildConfig.GOOGLE_PLAY)
                            showLocationRequest(activity)
                        else
                            onRequestLocationQ.launch(PERMISSION_FINE_LOCATION)
                    }
                }
            }
            else -> {
                when {
                    ActivityCompat.shouldShowRequestPermissionRationale(
                        activity, PERMISSION_COURSE_LOCATION
                    ) -> {
                        showLocationRequest(activity)
                    }
                    ContextCompat.checkSelfPermission(
                        activity, PERMISSION_COURSE_LOCATION
                    ) == PackageManager.PERMISSION_GRANTED -> {
                        requestBluetooth(activity)
                    }
                    else -> {
                        if (BuildConfig.GOOGLE_PLAY)
                            showLocationRequest(activity)
                        else
                            onRequestLocation.launch(PERMISSION_COURSE_LOCATION)
                    }
                }
            }
        }
    }

    private fun enableBluetoothAdapter(
        context: Context, adapter: BluetoothAdapter
    ) : BluetoothAdapter {
        return adapter.also {
            if (it.isEnabled) return it
            if (Version.isTiramisu) {
                AlertDialog.Builder(context)
                    .setMessage(R.string.tiramisu_bluetooth)
                    .setCancelable(false)
                    .setPositiveButton(R.string.proceed) { dialog: DialogInterface, _: Int ->
                        onRequestAdapter.launch(
                                try {
                                    Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
                                } catch (anf: ActivityNotFoundException) {
                                    Intent(Settings.ACTION_BLUETOOTH_SETTINGS)
                                })
                        dialog.dismiss()
                    }
                    .setNegativeButton(R.string.cancel) { dialog: DialogInterface, _: Int ->
                        dialog.dismiss()
                    }
                    .show()
                listener.onAdapterMissing()
            } else {
                try {
                    @Suppress("deprecation")
                    it.enable()
                } catch (se: SecurityException) { listener.onPermissionsFailed() }
            }
        }
    }

    @SuppressLint("MissingPermission")
    fun getBluetoothAdapter(context: Context?): BluetoothAdapter? {
        if (null == context) return null
        return try {
            mBluetoothAdapter?.let { enableBluetoothAdapter(context, it) }
                ?: enableBluetoothAdapter(context,
                    with (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager) {
                        adapter
                    }
                )
        } catch (ex: SecurityException) {
            Debug.warn(ex)
            null
        }
    }

    fun unregisterResultContracts() {
        onRequestLocationQ.unregister()
        onRequestBluetoothS.unregister()
        onRequestBluetooth.unregister()
        onRequestLocation.unregister()
        onRequestAdapter.unregister()
    }

    companion object {
        private const val PERMISSION_COURSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION
        private const val PERMISSION_FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION
        @RequiresApi(Build.VERSION_CODES.S)
        private val PERMISSIONS_BLUETOOTH = arrayOf(
            Manifest.permission.BLUETOOTH_SCAN,
            Manifest.permission.BLUETOOTH_CONNECT
        )
    }

    interface BluetoothListener {
        fun onPermissionsFailed()
        fun onAdapterMissing()
        fun onAdapterRestricted()
        fun onAdapterEnabled(adapter: BluetoothAdapter?)
    }
}